step-templates/octopus-create-octoterra-space-s3.json

```diff
@@ -0,0 +1,174 @@
+  "Description": "This step exposes the fields required to create the initial blank space serialized with [octoterra](https://github.com/OctopusSolutionsEngineering/OctopusTerraformExport) using Terraform. The new space is then populated using the `Octopus - Populate Octoterra Space (S3 Backend)` step.\n\nThis step configures a Terraform S3 backend.\n\nIt is recommended that this step be run with the `octopuslabs/terraform-workertools` worker image.",
+  "ActionType": "Octopus.TerraformApply",
+  "Version": 1,
+  "CommunityActionTemplateId": null,
+  "Packages": [
+    {
+      "Id": "093b1515-15a9-4446-8dc2-6297018a77e7",
+      "Name": "",
+      "PackageId": null,
+      "FeedId": null,
+      "AcquisitionLocation": "Server",
+      "Properties": {
+        "SelectionMode": "deferred",
+        "PackageParameterName": "OctoterraApply.Terraform.Package.Id"
+      }
+    }
+  ],
+  "Properties": {
+    "Octopus.Action.GoogleCloud.UseVMServiceAccount": "False",
+    "Octopus.Action.GoogleCloud.ImpersonateServiceAccount": "False",
+    "Octopus.Action.Terraform.GoogleCloudAccount": "True",
+    "Octopus.Action.Terraform.AzureAccount": "True",
+    "Octopus.Action.Terraform.ManagedAccount": "AWS",
+    "Octopus.Action.Terraform.AllowPluginDownloads": "True",
+    "Octopus.Action.Script.ScriptSource": "Package",
+    "Octopus.Action.Terraform.RunAutomaticFileSubstitution": "False",
+    "Octopus.Action.Terraform.PlanJsonOutput": "False",
+    "Octopus.Action.Terraform.Workspace": "#{OctoterraApply.Terraform.Workspace.Name}",
+    "Octopus.Action.Terraform.AdditionalInitParams": "-backend-config=\"bucket=#{OctoterraApply.AWS.S3.BucketName}\" -backend-config=\"region=#{OctoterraApply.AWS.S3.BucketRegion}\" -backend-config=\"key=#{OctoterraApply.AWS.S3.BucketKey}\" #{if OctoterraApply.Terraform.AdditionalInitParams}#{OctoterraApply.Terraform.AdditionalInitParams}#{/if}",
+    "Octopus.Action.Terraform.AdditionalActionParams": "-var=octopus_server=#{OctoterraApply.Octopus.ServerUrl} -var=octopus_apikey=#{OctoterraApply.Octopus.ApiKey} -var=octopus_space_name=#{OctoterraApply.Octopus.Space.NewName} -var=octopus_space_managers=#{OctoterraApply.Octopus.Space.Managers} #{if OctoterraApply.Terraform.AdditionalApplyParams}#{OctoterraApply.Terraform.AdditionalApplyParams}#{/if}",
+    "Octopus.Action.Package.DownloadOnTentacle": "False",
+    "Octopus.Action.RunOnServer": "true",
+    "Octopus.Action.AwsAccount.UseInstanceRole": "False",
+    "Octopus.Action.AwsAccount.Variable": "OctoterraApply.AWS.Account",
+    "Octopus.Action.Aws.AssumeRole": "False",
+    "Octopus.Action.Aws.Region": "#{OctoterraApply.AWS.S3.BucketRegion}",
+    "Octopus.Action.Terraform.TemplateDirectory": "space_creation",
+    "Octopus.Action.AzureAccount.Variable": "a",
+    "Octopus.Action.GoogleCloudAccount.Variable": "a"
+  },
+  "Parameters": [
+    {
+      "Id": "737a32e1-802f-4e15-a67d-cd271aef2c27",
+      "Name": "OctoterraApply.Octopus.Space.NewName",
```

Comment:
Not a major deal (i.e., I don't think it causes any problems as described [here](https://github.com/OctopusDeploy/Issues/issues/2126)) but noticed the prefix of the parameters here are the same as in the `Octopus - Populate Octoterra Space (S3 Backend)` step.

Reply:
I started down the road of making these parameters unique, but realised that it was going to be a lot of work for no real benefit. For example, these apply steps will need a few different variations to accommodate all the different Terraform backends. I've started with the S3 versions to support the EKS reference architecture step, but will need Azure blob storage version to match a future Azure web app RA step, and then one for google as well.

My understanding is that step templates should expose very specialised parameters to avoid potential clashes with project variables. But those parameters don't need to be unique between step templates - otherwise using the same template twice in a deployment process would also have issues. So I've reused the prefix `OctoterraApply` in multiple steps rather than inventing new variable names each time.

Reply:
Yeah that makes sense, thanks!